File: desktop-viewer/layout-inspector/presentation/src/test/kotlin/dev/agentperf/desktop/PaneLayoutTest.kt
package dev.agentperf.desktop

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class PaneLayoutTest {
    @Test
    fun `default hierarchy pane width is twenty five percent wider`() {
        assertEquals(375f, PaneWidths().hierarchy)
    }

    @Test
    fun `dragging separators changes only the adjacent side pane`() {
        val initial = PaneWidths()
        assertEquals(
            PaneWidths(hierarchy = 455f, properties = 300f),
            PaneLayout.dragHierarchy(initial, deltaDp = 80f, availableWidthDp = 1200f),
        )
        assertEquals(
            PaneWidths(hierarchy = 375f, properties = 260f),
            PaneLayout.dragProperties(initial, deltaDp = 40f, availableWidthDp = 1200f),
        )
    }

    @Test
    fun `dragging clamps side panes and preserves canvas minimum width`() {
        val initial = PaneWidths()
        assertEquals(PaneWidths(180f, 300f), PaneLayout.dragHierarchy(initial, -1000f, 1100f))
        assertEquals(PaneWidths(466f, 300f), PaneLayout.dragHierarchy(initial, 1000f, 1100f))
        assertEquals(PaneWidths(375f, 240f), PaneLayout.dragProperties(initial, 1000f, 1100f))
        assertEquals(PaneWidths(375f, 391f), PaneLayout.dragProperties(initial, -1000f, 1100f))
        val hierarchyMaximum = PaneLayout.dragHierarchy(initial, 1000f, 1100f)
        assertEquals(hierarchyMaximum, PaneLayout.dragHierarchy(hierarchyMaximum, 1000f, 1100f))
    }

    @Test
    fun `fitting remembered widths preserves canvas minimum after window shrinks`() {
        val fitted = PaneLayout.fit(
            widths = PaneWidths(hierarchy = 500f, properties = 500f),
            availableWidthDp = 1100f,
        )

        assertEquals(PaneWidths(hierarchy = 500f, properties = 266f), fitted)
        assertEquals(
            PaneLayout.CANVAS_MIN_WIDTH_DP,
            1100f -
                fitted.hierarchy -
                fitted.properties -
                PaneLayout.SPLITTER_WIDTH_DP * 2,
        )
    }
}

File: electron-viewer/apps/desktop/src/renderer/src/layout-inspector/pane-layout.test.ts
import { describe, expect, it } from 'vitest';
import {
  DEFAULT_PANE_WIDTHS,
  PANE_LAYOUT,
  dragHierarchyWidth,
  dragPropertiesWidth,
  fitPaneWidths,
} from './pane-layout';

/** Every assertion here is PaneLayoutTest.kt's, ported line for line. */
describe('pane layout', () => {
  it('default hierarchy pane width is twenty five percent wider', () => {
    expect(DEFAULT_PANE_WIDTHS.hierarchy).toBe(375);
  });

  it('dragging separators changes only the adjacent side pane', () => {
    expect(dragHierarchyWidth(DEFAULT_PANE_WIDTHS, 80, 1200)).toEqual({ hierarchy: 455, properties: 300 });
    expect(dragPropertiesWidth(DEFAULT_PANE_WIDTHS, 40, 1200)).toEqual({ hierarchy: 375, properties: 260 });
  });

  it('dragging clamps side panes and preserves canvas minimum width', () => {
    expect(dragHierarchyWidth(DEFAULT_PANE_WIDTHS, -1000, 1100)).toEqual({ hierarchy: 180, properties: 300 });
    expect(dragHierarchyWidth(DEFAULT_PANE_WIDTHS, 1000, 1100)).toEqual({ hierarchy: 466, properties: 300 });
    expect(dragPropertiesWidth(DEFAULT_PANE_WIDTHS, 1000, 1100)).toEqual({ hierarchy: 375, properties: 240 });
    expect(dragPropertiesWidth(DEFAULT_PANE_WIDTHS, -1000, 1100)).toEqual({ hierarchy: 375, properties: 391 });
    const hierarchyMaximum = dragHierarchyWidth(DEFAULT_PANE_WIDTHS, 1000, 1100);
    expect(dragHierarchyWidth(hierarchyMaximum, 1000, 1100)).toEqual(hierarchyMaximum);
  });

  it('fitting remembered widths preserves canvas minimum after window shrinks', () => {
    const fitted = fitPaneWidths({ hierarchy: 500, properties: 500 }, 1100);

    expect(fitted).toEqual({ hierarchy: 500, properties: 266 });
    expect(1100 - fitted.hierarchy - fitted.properties - PANE_LAYOUT.splitterWidth * PANE_LAYOUT.splitterCount).toBe(
      PANE_LAYOUT.canvasMinWidth,
    );
  });
});
